### PR TITLE
tss/rsa: key generation for threshold RSA (safe primes)

### PR DESCRIPTION
**Pull request by armfazh**
Includes a method that generates RSA keys where the modulus is the product of two safe prime numbers.

### CODE REVIEW

File: math/primes.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"crypto/rand"
+	"io"
+	"math/big"
+)
+
+// IsSafePrime reports whether p is (probably) a safe prime.
+// The prime p=2*q+1 is safe prime if both p and q are primes.
+// The n parameter is the number of checks for primality test.
```
Comment (bwesterb) [inline math/primes.go:11]: Good to add the caveat from `ProbablyPrime`:

```
ProbablyPrime is not suitable for judging primes that an adversary may
have crafted to fool the test.
````

File: math/primes.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+func IsSafePrime(p *big.Int, n int) bool {
```
Comment (bwesterb) [inline math/primes.go:12]: Why make `n` configurable? Perhaps best to just use 20, as [crypto/rand.Prime](https://pkg.go.dev/crypto/rand#Prime) does.

File: math/primes.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	pdiv2 := new(big.Int).Rsh(p, 1)
+	return p.ProbablyPrime(n) && pdiv2.ProbablyPrime(n)
+}
+
+// SafePrime returns a number of the given bit length that is a safe prime with high probability.
+// The number returned p=2*q+1 is a safe prime if both p and q are primes.
+// SafePrime will return error for any error returned by rand.Read or if bits < 2.
+func SafePrime(random io.Reader, bits int) (*big.Int, error) {
```
Comment (bwesterb) [inline math/primes.go:20]: I think it's faster to find a prime `q`, and then check whether `2q+1` is prime.

**Review by bwesterb — APPROVED** [review]
I left some suggestions.

File: math/primes.go (hunk 1 above)
Comment (armfazh) [reply math/primes.go:11]: added

File: math/primes.go (hunk 2 above)
Comment (armfazh) [reply math/primes.go:12]: removed the parameter.

File: math/primes.go (hunk 3 above)
Comment (armfazh) [reply math/primes.go:20]: true, changed

**Pull request merged**